src/Compilers/Core/Portable/Serialization/SerializationThreadPool.cs

```diff
@@ -0,0 +1,170 @@
+                            {
+                                if (item.action is Func<object?, object?> pts)
+                                {
+                                    item.tcs.SetResult(pts(item.state));
+                                }
+                                else
+                                {
+                                    item.tcs.SetResult(((Func<object?>)item.action)());
+                                }
+                            }
+                            catch (OperationCanceledException ex)
+                            {
+                                item.tcs.TrySetCanceled(ex.CancellationToken);
+                            }
+                            catch (Exception ex)
+                            {
+                                item.tcs.TrySetException(ex);
+                            }
+                            finally
+                            {
+                                item.tcs.SetResult(true);
+                            }
+                        }
+                    });
+                    t.IsBackground = true;
+                    t.Start();
+                }
+
+                static void enqueue((Delegate, object?, TaskCompletionSource<object?>) item)
+                {
+                    // Enqueue the work. If there are currently fewer threads waiting
+                    // for work than there are work items in the queue, create another
+                    // thread. This is a heuristic, in that we might end up creating
+                    // more threads than are truly needed, but this whole type is being
+                    // used to replace a previous solution where every work item created
+                    // its own thread, so this is an improvement regardless of any
+                    // such inefficiencies.
+                    lock (s_queue)
+                    {
+                        s_queue.Enqueue(item);
+
+                        if (s_queue.Count <= s_availableThreads)
+                        {
+                            Monitor.Pulse(s_queue);
+                            return;
+                        }
+                    }
+
+                    // No thread was currently available.  Create one.
+                    createThread();
+                }
+
+                static bool tryDequeue(out (Delegate action, object? state, TaskCompletionSource<object?> tcs) item)
+                {
+                    // Dequeues the next item if one is available. Before checking,
+                    // the available thread count is increased, so that enqueuers can
+                    // see how many threads are currently waiting, with the count
+                    // decreased after. Each time it waits, it'll wait for at most
+                    // the idle timeout before giving up.
+                    lock (s_queue)
+                    {
+                        s_availableThreads++;
```

Comment:
this part is still the part that i'm not understanding.  shouldn't the ++ And -- be done in the createThread part so that we know if we have a thread running that is available to process things?  having it be in tryDequeue is honestly confusing my brain.

Reply:
`s_availableThreads` is the number of threads waiting in `tryDequeue`.

Reply:
Ah. Perhaps another name might help here. But it's not a big deal. If this babe works for you, that's fine with me.

Reply:
I left the name, but I did update the comment on the field to call out `tryDequeue` explicitly.